src/actions/exchangeActions.js

```diff
@@ -263,49 +270,84 @@ export const addExchangeAllowanceIfNeededAction = ({
+
     // filter pending for current provider and asset match to override failed transactions
-    const allowances = _allowances.filter(({
-      provider: _provider,
-      fromAssetCode: _fromAssetCode,
-      toAssetCode: _toAssetCode,
-    }) => fromAssetCode !== _fromAssetCode && toAssetCode !== _toAssetCode && provider !== _provider);
+    const updatedAllowances = Object.keys(allowances).reduce((updated, allowancesAccountId) => {
+      if (senderAccountId !== allowancesAccountId) return updated;
 
-    allowances.push(allowance);
+      const accountAllowances = allowances[allowancesAccountId] || [];
+
+      const updatedAccountAllowances = accountAllowances
+        .filter(({
+          provider: _provider,
+          fromAssetCode: _fromAssetCode,
+          toAssetCode: _toAssetCode,
+        }) => fromAssetCode !== _fromAssetCode && toAssetCode !== _toAssetCode && provider !== _provider)
+        .concat(allowance);
+
+      return { ...updated, [allowancesAccountId]: updatedAccountAllowances };
+    }, {});
 
     dispatch({
-      type: ADD_EXCHANGE_ALLOWANCE,
-      payload: allowance,
+      type: ADD_ACCOUNT_EXCHANGE_ALLOWANCE,
+      payload: { allowance, accountId: senderAccountId },
     });
-    dispatch(saveDbAction('exchangeAllowances', { allowances }, true));
+
+    dispatch(saveDbAction('exchangeAllowances', { allowances: updatedAllowances }, true));
   };
 };
 
-export const enableExchangeAllowanceByHashAction = (transactionHash: string) => {
-  return async (dispatch: Dispatch, getState: GetState) => {
-    const { exchange: { data: { allowances: _allowances = [] } } } = getState();
-    const allowance = _allowances.find(
+export const enableExchangeAllowanceByHashAction = (transactionHash: string, senderAccountId: string) => {
+  return (dispatch: Dispatch, getState: GetState) => {
+    const allowances = allAccountsAllowancesSelector(getState());
+
+    const combinedAllowances = Object.keys(allowances).reduce((combined, accountId) => [
+      ...combined,
+      ...allowances[accountId],
+    ], []);
+
+    const allowance = combinedAllowances.find(
       ({ transactionHash: _transactionHash }) => _transactionHash === transactionHash,
     );
+
     if (!allowance) return;
+
     const updatedAllowance = {
       ...allowance,
       enabled: true,
     };
+
     dispatch({
-      type: UPDATE_EXCHANGE_ALLOWANCE,
-      payload: updatedAllowance,
+      type: UPDATE_ACCOUNT_EXCHANGE_ALLOWANCE,
+      payload: { accountId: senderAccountId, allowance: updatedAllowance },
     });
-    const allowances = _allowances
-      .filter(
-        ({ transactionHash: _transactionHash }) => _transactionHash !== transactionHash,
-      );
-    allowances.push(updatedAllowance);
-    dispatch(saveDbAction('exchangeAllowances', { allowances }, true));
+
+    const updatedAllowances = Object.keys(allowances).reduce((updated, allowancesAccountId) => {
+      if (senderAccountId !== allowancesAccountId) return updated;
```

Comment:
This part seems a bit weird. So for all entries in `allowances` map we are interested only in allowanes for `senderAccountId`. IMO it looks like there is no need to use `reduce` then, but just get the `allowances[senderAccountId]` and do the same thing as in `reduce` body.

WDYT?

Reply:
You're absolutely right, this ticket is little rush of moving existing pieces, changed, thanks, good catch 👍